pkg/lexer/lexer.go

```diff
@@ -1,5 +1,215 @@
 package lexer
 
+import (
+	"strings"
+	"unicode"
+)
+
 func Run(input string) string {
-	return "Lexing:\n" + input
+	l := New(input)
+	tok := l.Next()
+	var sb strings.Builder
+	for ; tok.Type != EOF; tok = l.Next() {
+		sb.WriteString(tok.String() + "\n")
+	}
+	sb.WriteString(tok.String() + "\n")
+	return sb.String()
+}
+
+type Lexer struct {
+	input []rune
+	r     rune // current rune under examination
+	pos   int  // current position in input (points to current rune)
+	line  int
+	col   int
+}
+
+func New(input string) *Lexer {
+	return &Lexer{input: []rune(input), pos: -1, line: 1}
+}
+
+func (l *Lexer) readRune() {
+	l.pos += 1
+	l.col += 1
+	if l.r == '\n' {
+		l.line += 1
+		l.col = 1
+	}
+	if l.pos >= len(l.input) {
+		l.r = 0
+		return
+	}
+	l.r = l.input[l.pos]
+}
+
+func (l *Lexer) skipWhitespace() {
+	for l.r == ' ' || l.r == '\t' || l.r == '\r' {
+		l.readRune()
+	}
+}
+
+func (l *Lexer) peekRune() rune {
+	if l.pos+1 >= len(l.input) {
+		return 0
+	}
+	return l.input[l.pos+1]
+}
+
+func (l *Lexer) Next() *Token {
+	l.readRune()
+	l.skipWhitespace()
+
+	tok := &Token{
+		Offset: l.pos,
+		Line:   l.line,
+		Col:    l.col,
+	}
+	switch l.r {
+	case '=':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(EQ)
+		}
+		return tok.SetType(ASSIGN)
+	case '+':
+		return tok.SetType(PLUS)
+	case '-':
+		return tok.SetType(MINUS)
+	case '!':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(NOT_EQ)
+		}
+		return tok.SetType(BANG)
+	case '/':
+		if l.peekRune() == '/' {
+			return tok.SetType(COMMENT).SetLiteral(l.readComment())
+		}
+		return tok.SetType(SLASH)
+	case '*':
+		return tok.SetType(ASTERISK)
+	case '<':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(LTEQ)
+		}
+		return tok.SetType(LT)
+	case '>':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(GTEQ)
+		}
+		return tok.SetType(GT)
+	case ':':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(DECLARE)
+		}
+		return tok.SetType(COLON)
+	case '{':
+		return tok.SetType(LCURLY)
+	case '}':
+		return tok.SetType(RCURLY)
+	case '(':
+		return tok.SetType(LPAREN)
+	case ')':
+		return tok.SetType(RPAREN)
+	case '[':
+		return tok.SetType(LBRACKET)
+	case ']':
+		return tok.SetType(RBRACKET)
+	case '\n':
+		return tok.SetType(NL)
+	case '.':
+		if l.peekRune() == '.' && l.peekRune2() == '.' {
+			l.readRune()
+			l.readRune()
+			return tok.SetType(DOT3)
+		}
+		return tok.SetType(DOT)
+	case '"':
+		literal := l.readString()
+		// EOF or NL before closing `"`
+		if l.r != '"' {
+			return tok.SetType(ILLEGAL).SetLiteral(`"`)
+		}
+		return tok.SetType(STRING_LIT).SetLiteral(literal)
+	case 0:
+		return tok.SetType(EOF)
+	}
+	if isLetter(l.r) {
+		lit := l.readIdent()
```

Comment:
nit: you call this var `literal` a few lines above. perhaps keep the two the same?

Reply:
done.